Feature/robot localization listener

rokusottervanger opened the pull request:
This adds a class that can extrapolate using buffered states (the EKF way), called RobotLocalizationEstimator; a wrapper (RosRobotLocalizationListener) that subscribes to a nav_msgs/Odometry topic and a geometry_msgs/AccelWithCovarianceStamped topic and adds them to its instance of the estimator; and a node that holds an instance of the Listener and also offers the state with covariance at a requested time via a service call.

Commit:
Add robot localization estimator

Commit:
Copy predict from EKF, not compiling

Commit:
Add test program for localization estimator

Commit:
Remove control action

Commit:
Add(robot localization estimator): method to set buffer capacity

Commit:
Add(robot localization listener): Ros dependent class to wrap robot localization estimator

Commit:
Fix ns of topics and callback registration

Commit:
Add conversion from odom and accel to estimator state

Commit:
Fix listener: conversion to tf instead of eigen

Commit:
Remove unnecessary input field in robot loc estimator

Commit:
Add conversion from odom and accel covariance to estimator covariance

Commit:
Add time stamp from odom to estimator state in robot loc listener

Commit:
Fix: naming of destructor and definition of odomAndAccelCallback

Commit:
Add: set state of estimator in callback

Commit:
Add(robot_localization_listener): getState methods

Commit:
Add(robot_localization_listener): docstrings in header

Commit:
Add simple test for inter- and extrapolation of state vector

Commit:
Fix(robot_localization_estimator): initialization of matrices

Commit:
getState is public now, also returns an ok bool

Commit:
Add: optional namespace argument to ros listener

Commit:
Update docstrings

Commit:
Doc(RobotLocalizationListener): Clearer comments and docstrings

Commit:
Different return value of getState when extrapolating into the past

Commit:
Remove(robot_localization_estimator): Unused, unclear method

Commit:
Add(test robot_localization_estimator): Proper testing of estimator

Commit:
Fix(ros_robot_localization_listener): Properly handle return value of estimator_.getState()

Commit:
Fix(ros_robot_localization_listener): If asked for state at rostime zero, return at rostime now

Commit:
Add(robot_localization_listener_node): node and service

Commit:
Fix(CmakeLists): robot_localization_estimator test

Commit:
Fix(test_robot_localization_estimator): some incorrect assertions

Commit:
Fix(robot_localization_estimator): add state with equal time stamp after previous

ayrton04 commented:
Looks like you're missing a dependency:

```
08:59:15 /tmp/catkin_workspace/src/robot_localization/src/ros_robot_localization_listener.cpp:35:41: fatal error: eigen_conversions/eigen_msg.h: No such file or directory
```

Commit:
Remove(ros_robot_localization_listener): Remove unnecessary includes

Commit:
Update copyright notices in added files

ayrton04 commented:
PR job failed because of the `tf/tf.h` include. This is due to the fact that the package uses only `tf2`. Would it be much work to change it? The datatypes are all the same.

Commit:
Fix(ros_robot_localization_listener): using tf2 instead of tf

Commit:
Add(ros_robot_localization_listener): frame_id in getState request (no implementation yet)

Commit:
Feat(ros_robot_localization_listener): getState for any robot-fixed frame (implementation)

Commit:
Fix(ros_robot_localization_listener): make sure state and covariance are STATE_SIZE, and some clearer variable names

Commit:
Fix(ros_robot_localization_listener): initialize return state and covariance at zeros

Commit:
Fix(ros_robot_localization_listener): Catch tf lookup exception

Commit:
Bug indices

Commit:
Bug flipped frame_ids

Commit:
Fix(ros_robot_localization_listener): transpose of rotation matrix in velocity transformation

Commit:
Add(test): Ros Robot Localization Listener test

rokusottervanger commented:
If all checks pass, this should be it. The only thing that is missing is the coordinate transformation of covariance matrices. This is relevant when the state of a different frame than the base frame is requested (line 255 in ros_robot_localization_listener.cpp). If anyone could help with that, that would be greatly appreciated.

ayrton04 commented:
Great, thanks! It'll take me some time to review. Re: rotating covariance matrices, see this:

https://github.com/cra-ros-pkg/robot_localization/blob/kinetic-devel/src/ros_filter.cpp#L2796

...and this:

http://robotics.stackexchange.com/questions/2556/how-to-rotate-covariance

ayrton04 commented on CMakeLists.txt:
```diff
@@ -15,7 +15,8 @@ find_package(catkin REQUIRED COMPONENTS
   tf2_geometry_msgs
-  tf2_ros)
+  tf2_ros
+  eigen_conversions)
```
Please alphabetize

ayrton04 commented on CMakeLists.txt:
```diff
@@ -40,6 +41,7 @@ add_service_files(
   FILES
   SetPose.srv
   SetDatum.srv
+  GetState.srv
```
Please alphabetize, and feel free to alphabetize `SetDatum` as well.

ayrton04 commented on CMakeLists.txt:
```diff
@@ -75,6 +78,7 @@ catkin_package(
     tf2
     tf2_geometry_msgs
     tf2_ros
+    eigen_conversions
```
Please alphabetize

ayrton04 commented on CMakeLists.txt:
```diff
@@ -109,11 +118,15 @@ target_link_libraries(filter_base filter_utilities ${catkin_LIBRARIES} ${EIGEN3_
 target_link_libraries(navsat_transform filter_utilities ros_filter_utilities ${catkin_LIBRARIES} ${EIGEN3_LIBRARIES})
 target_link_libraries(navsat_transform_node navsat_transform ${catkin_LIBRARIES})
 
+
```
Please remove

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+//!
+struct EstimatorState
+{
+    EstimatorState():
```
My goal is to make any new classes/files added to the package conform to [ROS coding standards](http://wiki.ros.org/CppStyleGuide). To that end, indentations should be two spaces.

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+
+  private:
+
+    boost::circular_buffer<EstimatorState> state_buffer_;
```
Sorry to be pedantic, but can we move the variables below the method declarations?

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+    //! @brief Keeps the state's Roll, Pitch, Yaw angles in the range
+    //! [-pi, pi]
+    //!
+    void wrapStateAngles(EstimatorState state) const
```
Nothing for you to change here, but it looks like this method should be added to one of the utility classes, since we define the same method elsewhere.

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+
+    //! @brief Returns the state at a given time
+    //!
+    //! Projects the current state and error matrices forward using a model of
```
This looks like copy-and-paste description code from the `predict` method.

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+    //!
+    //! @return buffer capacity
+    //!
+    unsigned int capacity() const;
```
Can we make this `getBufferCapacity` so it matches the set method?

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+    //!
+    //! @return current buffer size
+    //!
+    unsigned int size() const;
```
Same here: `getBufferSize()` is clearer. Also, it might be good to add a comment about what's different between capacity and size.

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+    //! @param[in] requested_time - time stamp to extrapolate to
+    //! @param[out] state_at_req_time - predicted state at requested time
+    //!
+    void interpolate(const EstimatorState& given_state_1, const EstimatorState& given_state_2, const double requested_time, EstimatorState& state_at_req_time) const;
```
This looks like it's a lot longer than 120 lines. I'd run roslint on this and correct whatever it says.

ayrton04 commented on include/robot_localization/robot_localization_estimator.h:
```diff
@@ -0,0 +1,196 @@
+    //! add this matrix (times deltaT) to the state estimate covariance
+    //! matrix. This models the fact that the prediction is never perfect.
+    //!
+    Eigen::MatrixXd processNoiseCovariance_;
```
It's new code, so I'd use the correct format of `process_noise_covariance_`. Also, I'm a bit confused as to why this variable is here. Is this used in extrapolation?

ayrton04 commented on include/robot_localization/ros_robot_localization_listener.h:
```diff
@@ -0,0 +1,162 @@
+//! RobotLocalizationEstimator. It exposes a getState method to
+//! offer the user the estimated state at a requested time. This
+//! class offers the option to run this listener without the need to
+//! to run a separate node. If you do wish to run this functionality
```
Handy

ayrton04 commented on include/robot_localization/ros_robot_localization_listener.h:
```diff
@@ -0,0 +1,162 @@
+    //! from the Robot Localization Estimator.
+    //!
+    //! @param[in] time - time of the requested state
+    //! @param[out] frame_id - state at the requested time
```
Fix this comment

ayrton04 commented on include/robot_localization/ros_robot_localization_listener.h:
```diff
@@ -0,0 +1,162 @@
+
+    //! @brief Subscriber to the acceleration state topic (output of a filter node)
+    //!
+    message_filters::Subscriber<geometry_msgs::AccelWithCovarianceStamped> accel_sub_;
```
Looks like an extra line and a missed indentation here

ayrton04 commented on package.xml:
```diff
@@ -28,6 +28,7 @@
   <depend>tf2</depend>
   <depend>tf2_geometry_msgs</depend>
   <depend>tf2_ros</depend>
+  <depend>eigen_conversions</depend>
```
Please alphabetize

ayrton04 commented on src/robot_localization_estimator.cpp:
```diff
@@ -0,0 +1,330 @@
+  return state_buffer_.size();
+}
+
+void RobotLocalizationEstimator::extrapolate(const EstimatorState& boundary_state,
```
This is a lot of repeat code from the EKF. Any reason we can't just instantiate an EKF, manually set the process noise in the constructor, set the state and estimate error covariance here, and call predict, then getState? Also, we have two different filters, and I am working on 2D versions of each of them. It might be better to use a parameter that specifies what extrapolation model to use (i.e., EKF or UKF). You can then have a `FilterBase *filter_` member variable that you instantiate.

rokusottervanger replied:
This was fixed in 65d256cd4b297d430bb3c540a9eb907e3218d7bc and f24a843aad897e60282a1c3c4862af8743434bb9

ayrton04 commented on src/robot_localization_estimator.cpp:
```diff
@@ -0,0 +1,330 @@
+   * But as the state after the requested time is also known, we may want to perform
+   * interpolation between states.
+   */
+  extrapolate(given_state_1, requested_time, state_at_req_time);
```
I think this method *does* need to interpolate with the states on either side of the requested time. Otherwise, your requested state data is subject to the EKF/UKF motion model, which won't be accurate for every vehicle. Besides, you have the state data, might as well use it.

rokusottervanger replied:
I found a method for blending the covariance matrix [here](http://www.agi.com/downloads/resources/white-papers/CovarianceInterpolationFinal.pdf). But so far I haven't been able to find a good way for blending/interpolation of the Euler angles in the state representation, taking into account the wrapped state angles and velocity constriants on both sides. Do you have suggestions for this?

ayrton04 commented on src/robot_localization_listener_node.cpp:
```diff
@@ -0,0 +1,91 @@
+public:
+  RobotLocalizationListenerNode()
+  {
+    service_ = n_.advertiseService("get_robot_localization_state", &RobotLocalizationListenerNode::getStateCallback, this);
```
We can probably remove the node name from the service call. Maybe just `get_state`?

ayrton04 commented on src/robot_localization_listener_node.cpp:
```diff
@@ -0,0 +1,91 @@
+                        robot_localization::GetState::Response &res)
+  {
+    Eigen::VectorXd state(RobotLocalization::STATE_SIZE);
+    Eigen::MatrixXd covariance(RobotLocalization::STATE_SIZE,RobotLocalization::STATE_SIZE);
```
Maybe add a `using RobotLocalization::STATE_SIZE` or something? Seems awkward to have to use the qualifier everywhere.

ayrton04 commented on src/robot_localization_listener_node.cpp:
```diff
@@ -0,0 +1,91 @@
+      res.covariance[i] = (*(covariance.data() + i));
+    }
+
+    ROS_INFO("Robot Localization Listener Node: Listener responded with state and covariance at the requested time.");
```
Will this get spammy if users are making frequent calls to this service? Might be better for a debug message.

ayrton04 commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -0,0 +1,278 @@
+
+    estimator_.setBufferCapacity(buffer_size);
+
+    sync_.registerCallback(&RosRobotLocalizationListener::odomAndAccelCallback, this);
```
Indentation

ayrton04 commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -0,0 +1,278 @@
+    state.state(StateMemberZ) = odom.pose.pose.position.z;
+
+    // Pose: Orientation
+    tf2::Quaternion orientation_quat;
```
I definitely have a `quatToRpy` method in one of the utility headers.

ayrton04 commented on src/robot_localization_estimator.cpp:
```diff
@@ -0,0 +1,330 @@
+  // If there's nothing in the buffer, there's nothing to give.
+  if ( state_buffer_.size() == 0 )
+  {
+    return -1;
```
No magic numbers, please. Add an enumeration for return codes if you really need them, or just make this a bool and don't worry about passing on the reason for failure.

ayrton04 commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -0,0 +1,278 @@
+      return false;
+    }
+
+    if ( estimator_.getState(time, estimator_state) == -2 )
```
Yeah, make some enumerated return codes.

ayrton04 commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -0,0 +1,278 @@
+    state(StateMemberYaw)   = ypr[0];
+
+    // Now let's calculate the twist of the target frame
+    Twist base_vel, target_vel_base;
```
One variable declaration per line, please. Most of this will be caught by roslint or just from reading over the style guide.

ayrton04 commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -0,0 +1,278 @@
+                                       estimator_state.state(StateMemberVyaw));
+
+    Eigen::Vector3d target_position_base = target_pose_base.translation();
+    target_vel_base.linear = base_vel.linear + base_vel.angular.cross(target_position_base);
```
I need to work this out on paper, but why are we not rotating the linear velocity before applying the cross product of the offset with the angular velocity? Here's how I'm handing it in the rest of the filter:

https://github.com/cra-ros-pkg/robot_localization/blob/kinetic-devel/src/ros_filter.cpp#L2769

ayrton04 replied:
Also, see this in the TF library:

https://github.com/ros/geometry/blob/4421e22243a38ba6a130494698b33117d2995a18/tf/src/tf.cpp#L317

rokusottervanger replied:
No rotation is necessary, because all vectors here are expressed in the base-fixed frame.

ayrton04 reviewed (changes requested):
First, thanks for the PR! There are some things I'd request that you change. I think you could re-use code from the EKF/UKF for handling extrapolation, and it would be good to do (linear) interpolation between adjacent states rather than extrapolating from the previous state. Other than that, it's just  matter of a few coding standard violations. Thanks again!

rokusottervanger commented:
Thanks for the review! I will process your comments as soon as possible, but it may be a week or two before I have the time.

Commit:
Clean(localization listener): conform to style guide

Commit:
Refactor(robot_localization_estimator): WIP: Use existing filter implementations for extrapolation

Commit:
test(robot_localization_estimator): fix estimator construction

Commit:
chore(ros robot localization listener): add some comments for better explanation

Commit:
feat(localization estimator): Add case for when requested time is equal to a buffered time

Commit:
fix(localization estimator): require process noise covariance at construction

Commit:
test(localization estimator): more elaborate test

Commit:
fix(localization estimator): use getState instead of getPredictedState

Commit:
fix(ros robot localization listener): looks for process noise param in namespace of filter node that it listens to

ayrton04 commented:
I take it this is still in progress, right?

rokusottervanger commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -31,248 +31,359 @@
+  state(StateMemberVyaw) = target_velocity.angular(2);
+
+  // TODO: transform covariance
+  covariance = estimator_state.covariance;
```
This covariance transformation is still a todo. A choice needs to be made though, because the rotational uncertainty in combination with translation implies a non-Gaussian uncertainty after transformation, so some approximation is needed. Do you have a suggestion, @ayrton04?

rokusottervanger replied:
Fixed in 3bcb157b52e4420891528ad23312e280c1e57b0b. The comment still holds, but it is now solved the same way as in [navsat_transform](https://github.com/cra-ros-pkg/robot_localization/blob/kinetic-devel/src/navsat_transform.cpp#L573)

rokusottervanger commented on src/robot_localization_estimator.cpp:
```diff
@@ -319,9 +202,8 @@ void RobotLocalizationEstimator::interpolate(const EstimatorState& given_state_1
-   * TODO: Right now, we only extrapolate from the last known state before the requested time.
-   * But as the state after the requested time is also known, we may want to perform
-   * interpolation between states.
+   * TODO: Right now, we only extrapolate from the last known state before the requested time. But as the state after
```
This todo is still open. The state interpolation can be done using state estimates on both sides of the requested time as constraints, but I have no knowledge about interpolation of the covariance matrices. By the way, interpolation does not take into account the system model from the filter, so extrapolation from the nearest known time might actually be a nicer option.

Commit:
feature(robot_localization_listener): Option to specify a world frame_id

If you would like to obtain the robot state in "utm" frame for example.
Also added the covariance transformation.

rokusottervanger commented on src/ros_robot_localization_listener.cpp:
```diff
@@ -0,0 +1,442 @@
+                                            Eigen::MatrixXd& covariance,
+                                            std::string world_frame_id)
+{
+  EstimatorState estimator_state;
```
This is going to need some check for the nature of the requested transform. Right now, requesting the state of "world" would give a reasonable result without complaining, but it would use the robot state to transform between odom and base, and then tf to transform between base and utm. We can either choose to throw an error if base is not a parent of the requested frame id, or only transform using tf if this is the case.

Commit:
fix(ros_robot_localization_listener): Check for validity of using robot localization listener for requested transform

Commit:
fix(robot_localization): add dependency on yaml-cpp

Commit:
refactor(findAncestor): Renamed functions and some cleanup

Naming was a little bit weird and I also added an additional method that
consumes the transformation buffer instead of the YAML.

Commit:
fix(lookupTransform): Use transformException

A lookUpException gives a runtime error if there is not connectivity
what usually happens on initialization.

Commit:
feature(initialization): Initialization update for frames and parameters

In the constructor, the robot localization listener waits for incoming
message to initialize properly. This enables the exposition of the
getFrame id methods. Also added an additional parameter_namespace
parameter to refer to the robot localization instance so that the same
filter parameters can be used, this value defaults to ~ which implies
that we search in the local namespace.

Commit:
fix(tests): Fixed tests for robot localization listener

Commit:
refactor(robot_localization_listener): Remap arg to robot_localization instance

The parameter namespace of the listener is now by default 'robot_localization',
this parameter can be remapped to the running robot localization listener to
obtain the same parameters as the running robot localization instance.

Commit:
feature(localization_listener): Added caching to ancestor lookup

ayrton04 commented:
Hi! Sorry for the lapse in comms. What's the status on this?

Commit:
Merge branch 'kinetic-devel' into feature/robot-localization-listener

rokusottervanger commented:
I will merge the upstream and resolve any conflicts. 
Last unresolved comment is about interpolation between states and corresponding covariance matrices. However, I currently don't know of a decent way to do both interpolation/blending between states and covariance matrices **and** take into account the system model. AFAICT, A prediction based on the previous state will suffice for all practical cases, and I would like to leave this feature open for future development.

ayrton04 commented:
Sounds good.

ayrton04 commented:
Are you happy for me to merge this at this point?

rokusottervanger commented:
Yes, that would be great.

On Aug 9, 2017 19:13, "Tom Moore" <notifications@github.com> wrote:

> Are you happy for me to merge this at this point?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cra-ros-pkg/robot_localization/pull/324#issuecomment-321237528>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIlLByZf-f6JZ2PeZaazH_iN2XrgKly5ks5sWaJlgaJpZM4Kxidf>
> .
>

ayrton04 commented:
TY for this PR!
